File: src/main/java/com/mohe/spring/service/InternalPlaceIngestService.java
package com.mohe.spring.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.mohe.spring.controller.InternalBatchController;
import com.mohe.spring.entity.Place;
import com.mohe.spring.repository.PlaceRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

@Service
public class InternalPlaceIngestService {

    private static final Logger log = LoggerFactory.getLogger(InternalPlaceIngestService.class);
    private final PlaceRepository placeRepository;
    private final ObjectMapper objectMapper;
    
    public InternalPlaceIngestService(PlaceRepository placeRepository, ObjectMapper objectMapper) {
        this.placeRepository = placeRepository;
        this.objectMapper = objectMapper;
    }
    
    @Transactional
    public InternalBatchController.InternalPlaceIngestResponse ingestPlaces(List<InternalBatchController.InternalPlaceIngestRequest> requests) {
        log.info("Starting place ingestion for {} requests", requests.size());
        int processedCount = 0;
        int insertedCount = 0;
        int updatedCount = 0;
        int skippedCount = 0;
        int errorCount = 0;
        List<String> errors = new ArrayList<>();
        
        for (InternalBatchController.InternalPlaceIngestRequest request : requests) {
            try {
                processedCount++;
                log.debug("Processing place {}/{}: {}", processedCount, requests.size(), request.getName());
                log.info("NaverID: " + request.getNaverPlaceId() + ", GoogleID: " + request.getGooglePlaceId());
                
                // Check if place already exists by naverPlaceId or googlePlaceId
                Optional<Place> existingPlace = findExistingPlace(request);
                
                if (existingPlace.isPresent()) {
                    // Update existing place
                    log.info("UPDATING existing place: " + request.getName() + " (DB ID: " + existingPlace.get().getId() + ")");
                    Place place = existingPlace.get();
                    updatePlaceFromRequest(place, request);
                    Place saved = placeRepository.save(place);
                    log.info("Successfully updated place with ID: " + saved.getId());
                    updatedCount++;
                } else {
                    // Create new place
                    log.info("CREATING new place: " + request.getName());
                    Place newPlace = createPlaceFromRequest(request);
                    Place saved = placeRepository.save(newPlace);
                    log.info("Successfully created NEW place with ID: " + saved.getId());
                    insertedCount++;
                }
                
            } catch (org.springframework.dao.DataIntegrityViolationException e) {
                log.info("CONSTRAINT VIOLATION for place: " + request.getName());
                log.info("NaverID: " + request.getNaverPlaceId() + ", GoogleID: " + request.getGooglePlaceId());
                log.info("Error message: " + e.getMessage());
                
                // Try to find existing place one more time with more debug info
                log.info("Attempting secondary lookup for constraint violation...");
                Optional<Place> debugLookup = findExistingPlace(request);
                if (debugLookup.isPresent()) {
                    log.info("FOUND place in secondary lookup - this indicates a race condition or caching issue");
                } else {
                    log.info("Still no place found in secondary lookup - possible data issue");
                }
                
                skippedCount++;
            } catch (Exception e) {
                log.info("ERROR processing place: " + request.getName() + " - " + e.getMessage());
                e.printStackTrace();
                errorCount++;
                errors.add("Failed to process place '" + request.getName() + "': " + e.getMessage());
            }
        }
        
        return new InternalBatchController.InternalPlaceIngestResponse(
            processedCount,
            insertedCount,
            updatedCount,
            skippedCount,
            errorCount,
            0, // keywordGeneratedCount - TODO: implement keyword processing
            0, // imagesFetchedCount - TODO: implement image fetching
            errors
        );
    }
    
    private Optional<Place> findExistingPlace(InternalBatchController.InternalPlaceIngestRequest request) {
        log.info("Checking for existing place - NaverID: " + request.getNaverPlaceId() + ", GoogleID: " + request.getGooglePlaceId());
        
        if (request.getNaverPlaceId() != null && !request.getNaverPlaceId().trim().isEmpty()) {
            Optional<Place> byNaverId = placeRepository.findByNaverPlaceId(request.getNaverPlaceId().trim());
            if (byNaverId.isPresent()) {
                log.info("Found existing place by NaverID: " + byNaverId.get().getId());
                return byNaverId;
            }
        }
        
        if (request.getGooglePlaceId() != null && !request.getGooglePlaceId().trim().isEmpty()) {
            Optional<Place> byGoogleId = placeRepository.findByGooglePlaceId(request.getGooglePlaceId().trim());
            if (byGoogleId.isPresent()) {
                log.info("Found existing place by GoogleID: " + byGoogleId.get().getId());
                return byGoogleId;
            }
        }
        
        log.info("No existing place found for: " + request.getName());
        return Optional.empty();
    }
    
    private Place createPlaceFromRequest(InternalBatchController.InternalPlaceIngestRequest request) {
        Place place = new Place(request.getName());
        updatePlaceFromRequest(place, request);
        return place;
    }
    
    private void updatePlaceFromRequest(Place place, InternalBatchController.InternalPlaceIngestRequest request) {
        try {
            place.setName(request.getName());
            place.setTitle(request.getName()); // For backward compatibility
            place.setDescription(request.getDescription());
            place.setCategory(request.getCategory());
            place.setAddress(request.getAddress());
            place.setRoadAddress(request.getRoadAddress());
            place.setLatitude(request.getLatitude());
            place.setLongitude(request.getLongitude());
            place.setPhone(request.getPhone());
            place.setWebsiteUrl(request.getWebsiteUrl());
            
            if (request.getRating() != null) {
                place.setRating(java.math.BigDecimal.valueOf(request.getRating()));
            }
            
            place.setUserRatingsTotal(request.getUserRatingsTotal());
            
            if (request.getPriceLevel() != null) {
                place.setPriceLevel(request.getPriceLevel().shortValue());
            }
            
            place.setNaverPlaceId(request.getNaverPlaceId());
            place.setGooglePlaceId(request.getGooglePlaceId());
            // Add image to gallery if provided
            if (request.getImageUrl() != null && !request.getImageUrl().trim().isEmpty()) {
                List<String> gallery = place.getGallery();
                if (gallery == null) {
                    gallery = new ArrayList<>();
                    place.setGallery(gallery);
                }
                if (!gallery.contains(request.getImageUrl().trim())) {
                    gallery.add(request.getImageUrl().trim());
                }
            }
            
            if (request.getTypes() != null) {
                place.setTypes(request.getTypes());
            }
            
            // Parse JSON fields
            if (request.getOpeningHours() != null) {
                JsonNode openingHoursJson = objectMapper.readTree(request.getOpeningHours());
                place.setOpeningHours(openingHoursJson);
            }
            
            if (request.getSourceFlags() != null) {
                JsonNode sourceFlagsJson = objectMapper.valueToTree(request.getSourceFlags());
                place.setSourceFlags(sourceFlagsJson);
            }
            
            // Store keyword vector as JSON string
            if (request.getKeywordVector() != null && !request.getKeywordVector().isEmpty()) {
                String vectorJson = objectMapper.writeValueAsString(request.getKeywordVector());
                place.setKeywordVector(vectorJson);
            }
            
            place.setUpdatedAt(java.time.LocalDateTime.now());
            
        } catch (Exception e) {
            throw new RuntimeException("Failed to update place from request", e);
        }
    }
    
    public Object triggerDataIngestion() {
        // TODO: Implement data ingestion trigger
        throw new UnsupportedOperationException("Data ingestion trigger not yet implemented");
    }
    
    public Object getIngestionStatus() {
        // TODO: Implement ingestion status retrieval
        throw new UnsupportedOperationException("Ingestion status retrieval not yet implemented");
    }
    
    public Object triggerSpecificIngestion(String category, int limit) {
        // TODO: Implement specific category ingestion
        throw new UnsupportedOperationException("Specific category ingestion not yet implemented");
    }
}

File: src/main/kotlin/com/mohe/spring/service/InternalPlaceIngestService.kt
package com.mohe.spring.service

import com.mohe.spring.controller.InternalPlaceIngestRequest
import com.mohe.spring.controller.InternalPlaceIngestResponse
import com.mohe.spring.entity.Place
import com.mohe.spring.repository.PlaceRepository
import com.fasterxml.jackson.databind.ObjectMapper
import org.slf4j.LoggerFactory
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.math.BigDecimal
import java.time.LocalDate
import java.time.OffsetDateTime
import java.util.concurrent.CompletableFuture

/**
 * Service for handling internal place ingestion from MoheBatch
 * Integrates with image fetching from Google Places
 */
@Service
@Transactional
class InternalPlaceIngestService(
    private val placeRepository: PlaceRepository,
    private val googlePlacesImageService: GooglePlacesImageService,
    private val objectMapper: ObjectMapper
) {

    private val logger = LoggerFactory.getLogger(InternalPlaceIngestService::class.java)

    /**
     * Ingest places from batch process and trigger image fetching
     */
    fun ingestPlaces(requests: List<InternalPlaceIngestRequest>): InternalPlaceIngestResponse {
        logger.info("Processing ${requests.size} places from batch ingestion")
        
        var insertedCount = 0
        var updatedCount = 0
        var skippedCount = 0
        var errorCount = 0
        var keywordGeneratedCount = 0
        var imagesFetchedCount = 0
        val errors = mutableListOf<String>()
        
        val imageFetchFutures = mutableListOf<CompletableFuture<Int>>()

        for (request in requests) {
            try {
                // Check if place already exists
                val existingPlace = findExistingPlace(request)
                
                val place = if (existingPlace != null) {
                    // Update existing place
                    val updatedPlace = updateExistingPlace(existingPlace, request)
                    placeRepository.save(updatedPlace)
                    updatedCount++
                    updatedPlace
                } else {
                    // Create new place
                    val newPlace = createPlaceFromRequest(request)
                    if (newPlace.shouldBeRecommended()) {
                        val savedPlace = placeRepository.save(newPlace)
                        insertedCount++
                        savedPlace
                    } else {
                        logger.debug("Skipping place ${request.name} - doesn't meet recommendation criteria")
                        skippedCount++
                        continue
                    }
                }

                // Count keywords if vector is present
                if (request.keywordVector.isNotEmpty()) {
                    keywordGeneratedCount++
                }

                // Trigger asynchronous image fetching for the place
                if (place.id != null) {
                    try {
                        val imageFuture = googlePlacesImageService.fetchImagesForPlace(place)
                        imageFetchFutures.add(imageFuture)
                        
                        // Don't wait for completion - just track the future
                        imageFuture.thenAccept { fetchedCount ->
                            logger.debug("Fetched $fetchedCount images for place: ${place.name}")
                        }.exceptionally { ex ->
                            logger.warn("Failed to fetch images for place ${place.name}: ${ex.message}")
                            null
                        }
                    } catch (ex: Exception) {
                        logger.warn("Failed to trigger image fetching for place ${place.name}: ${ex.message}")
                        errors.add("Image fetching failed for ${place.name}: ${ex.message}")
                    }
                }

            } catch (ex: Exception) {
                logger.error("Failed to process place ${request.name}: ${ex.message}", ex)
                errorCount++
                errors.add("Failed to process ${request.name}: ${ex.message}")
            }
        }

        // Wait briefly for some image fetches to complete (non-blocking approach)
        try {
            CompletableFuture.allOf(*imageFetchFutures.toTypedArray())
                .orTimeout(5, java.util.concurrent.TimeUnit.SECONDS) // Don't wait too long
                .thenRun {
                    imagesFetchedCount = imageFetchFutures.sumOf { 
                        try { it.get() } catch (e: Exception) { 0 }
                    }
                    logger.info("Batch ingestion completed with $imagesFetchedCount total images fetched")
                }
                .exceptionally { ex ->
                    logger.debug("Some image fetches still in progress: ${ex.message}")
                    null
                }
        } catch (ex: Exception) {
            logger.debug("Image fetching in progress asynchronously")
        }

        logger.info("Batch ingestion completed: $insertedCount inserted, $updatedCount updated, $skippedCount skipped, $errorCount errors")
        
        return InternalPlaceIngestResponse(
            processedCount = requests.size,
            insertedCount = insertedCount,
            updatedCount = updatedCount,
            skippedCount = skippedCount,
            errorCount = errorCount,
            keywordGeneratedCount = keywordGeneratedCount,
            imagesFetchedCount = imagesFetchedCount,
            errors = errors
        )
    }

    /**
     * Find existing place by Naver or Google ID
     */
    private fun findExistingPlace(request: InternalPlaceIngestRequest): Place? {
        // Try to find by Naver Place ID first
        placeRepository.findByNaverPlaceId(request.naverPlaceId)?.let { return it }
        
        // Try to find by Google Place ID if available
        request.googlePlaceId?.let { googleId ->
            placeRepository.findByGooglePlaceId(googleId)?.let { return it }
        }
        
        // Try to find similar place by name and location
        return placeRepository.findSimilarPlace(
            request.name,
            request.latitude,
            request.longitude,
            BigDecimal("0.001") // ~100m radius
        )
    }

    /**
     * Update existing place with new data
     */
    private fun updateExistingPlace(existingPlace: Place, request: InternalPlaceIngestRequest): Place {
        return existingPlace.copy(
            title = request.name,
            name = request.name,
            description = request.description,
            category = request.category,
            address = request.address,
            roadAddress = request.roadAddress,
            latitude = request.latitude,
            longitude = request.longitude,
            phone = request.phone,
            websiteUrl = request.websiteUrl,
            rating = request.rating?.let { BigDecimal(it.toString()) } ?: existingPlace.rating,
            userRatingsTotal = request.userRatingsTotal,
            priceLevel = request.priceLevel?.toShort(),
            types = request.types,
            openingHours = parseOpeningHours(request.openingHours),
            imageUrl = request.imageUrl ?: existingPlace.imageUrl,
            googlePlaceId = request.googlePlaceId ?: existingPlace.googlePlaceId,
            updatedAt = java.time.LocalDateTime.now(),
            lastRatingCheck = OffsetDateTime.now(),
            shouldRecheckRating = false // Just updated, no need to recheck yet
        )
    }

    /**
     * Create new place from request
     */
    private fun createPlaceFromRequest(request: InternalPlaceIngestRequest): Place {
        val rating = request.rating ?: 0.0
        val openedDate = extractOpeningDate(request)
        val isNewPlace = openedDate?.isAfter(LocalDate.now().minusMonths(6)) ?: true

        return Place(
            name = request.name,
            title = request.name,
            description = request.description,
            category = request.category,
            address = request.address,
            roadAddress = request.roadAddress,
            location = request.address,
            latitude = request.latitude,
            longitude = request.longitude,
            phone = request.phone,
            websiteUrl = request.websiteUrl,
            rating = BigDecimal(rating.toString()),
            reviewCount = request.userRatingsTotal ?: 0,
            userRatingsTotal = request.userRatingsTotal,
            priceLevel = request.priceLevel?.toShort(),
            types = request.types,
            openingHours = parseOpeningHours(request.openingHours),
            imageUrl = request.imageUrl,
            naverPlaceId = request.naverPlaceId,
            googlePlaceId = request.googlePlaceId,
            openedDate = openedDate,
            firstSeenAt = OffsetDateTime.now(),
            isNewPlace = isNewPlace,
            shouldRecheckRating = !isNewPlace, // Old places should be rechecked
            createdAt = OffsetDateTime.now(),
            updatedAt = java.time.LocalDateTime.now()
        )
    }

    /**
     * Parse opening hours JSON string
     */
    private fun parseOpeningHours(openingHoursJson: String?): com.fasterxml.jackson.databind.JsonNode? {
        if (openingHoursJson.isNullOrBlank()) return null
        
        return try {
            objectMapper.readTree(openingHoursJson)
        } catch (ex: Exception) {
            logger.debug("Failed to parse opening hours JSON: ${ex.message}")
            null
        }
    }

    /**
     * Extract opening date from request data
     */
    private fun extractOpeningDate(request: InternalPlaceIngestRequest): LocalDate? {
        // This could be enhanced to parse opening date from Google data
        // For now, return null (unknown opening date)
        return null
    }
}